### PR TITLE
[AArch64][GlobalISel] Avoid generating inserts for undefs when selecting G_BUILD_VECTOR

### PR DESCRIPTION
**Pull request by dc03-work**
It is safe to ignore undef values when selecting G_BUILD_VECTOR as undef values choose random registers for copying values from.

### CODE REVIEW

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp [hunk 1]
```diff
@@ -5963,12 +5966,20 @@ bool AArch64InstructionSelector::selectBuildVector(MachineInstr &I,
     MachineOperand &RegOp = I.getOperand(1);
     RegOp.setReg(Reg);
     RBI.constrainGenericRegister(DstReg, *RC, MRI);
-  } else {
+  } else if (PrevMI) {
```
Comment (madhur13490) [inline llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:5969]: Please explain in comments why `else if(PrevMi)` is needed.

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp [hunk 2]
```diff
@@ -5963,12 +5966,20 @@ bool AArch64InstructionSelector::selectBuildVector(MachineInstr &I,
...
     // We don't need a subregister copy. Save a copy by re-using the
     // destination register on the final insert.
-    assert(PrevMI && "PrevMI was null?");
     PrevMI->getOperand(0).setReg(I.getOperand(0).getReg());
     constrainSelectedInstRegOperands(*PrevMI, TII, TRI, RBI);
+  } else {
+    // All the operands (other than the first one) to the G_BUILD_VECTOR were
+    // undef, so PrevMI is nullptr. Emit a copy from the vector made from the
+    // first operand to the destination register.
```
Comment (madhur13490) [inline llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:5977]: "Emit a copy from the vector made from the
    // first operand to the destination register."

This is a bit difficult to read. Please simplify.

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp [hunk 3]
```diff
@@ -5963,12 +5966,20 @@ bool AArch64InstructionSelector::selectBuildVector(MachineInstr &I,
...
     RBI.constrainGenericRegister(DstReg, *RC, MRI);
-  } else {
+  } else if (PrevMI) {
     // We don't need a subregister copy. Save a copy by re-using the
     // destination register on the final insert.
-    assert(PrevMI && "PrevMI was null?");
     PrevMI->getOperand(0).setReg(I.getOperand(0).getReg());
     constrainSelectedInstRegOperands(*PrevMI, TII, TRI, RBI);
+  } else {
+    // All the operands (other than the first one) to the G_BUILD_VECTOR were
```
Comment (madhur13490) [inline llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:5975]: other than the first one -> except first one?

**Review by davemgreen — COMMENTED** [review]
Thanks this looks like a nice improvement.

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp [hunk 4]
```diff
@@ -5963,12 +5966,21 @@ bool AArch64InstructionSelector::selectBuildVector(MachineInstr &I,
...
-    // We don't need a subregister copy. Save a copy by re-using the
-    // destination register on the final insert.
-    assert(PrevMI && "PrevMI was null?");
+  } else if (PrevMI) {
+    // PrevMI is not nullptr, so we have generated inserts and don't need a
+    // subregister copy. Save a copy by re-using the destination register on the
+    // final insert.
     PrevMI->getOperand(0).setReg(I.getOperand(0).getReg());
     constrainSelectedInstRegOperands(*PrevMI, TII, TRI, RBI);
+  } else {
```
[the remark this answers is not in the document]
Comment (dc03-work) [reply llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:5975]: I did try that at first, however that was giving a crash:
```
LLVM ERROR: VReg has no regclass after selection: %bv:fpr(<4 x s32>) = INSERT_SUBREG %3:fpr128(tied-def 0), %val:gpr32all(s32), %subreg.ssub (in function: undef_elts_different_regbanks)
```
in `llvm/test/CodeGen/AArch64/GlobalISel/select-build-vector.mir`. I think it is best to try and elide the copy wherever possible.

[the remark this answers is not in the document]
Comment (davemgreen) [reply llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:5984]: I see. Which one? Setting PrevMI = ScalarToVec? It sound like the output of the variable could not be constrained as INSERT_SUBREG is a generic operation, would it work better if it was constrained to fpr128?

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp [hunk 5]
```diff
@@ -5934,13 +5934,16 @@ bool AArch64InstructionSelector::selectBuildVector(MachineInstr &I,
...
+  MachineInstr *PrevMI = ScalarToVec;
   for (unsigned i = 2, e = DstSize / EltSize + 1; i < e; ++i) {
     // Note that if we don't do a subregister copy, we can end up making an
     // extra register.
-    PrevMI = &*emitLaneInsert(std::nullopt, DstVec, I.getOperand(i).getReg(),
-                              i - 1, RB, MIB);
-    DstVec = PrevMI->getOperand(0).getReg();
+    Register OpReg = I.getOperand(i).getReg();
+    // Do not emit inserts for undefs
+    if (!getOpcodeDef(TargetOpcode::G_IMPLICIT_DEF, OpReg, MRI)) {
```
Comment (aemerson) [inline llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:5943]: You can do the shorter `!isa<GImplicitDef>(OpReg, MRI)`

**Review by aemerson — APPROVED** [review]
Thanks, LGTM with a minor suggestion.

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp (hunk 5 above)
Comment (aemerson) [reply llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:5943]: Sorry I mistyped, I've updated the previous comment with the right function name.

**Review by davemgreen — APPROVED** [review]
Thanks. LGTM too.

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp (hunk 1 above)
Comment (dc03-work) [reply llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp]: Done.

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp (hunk 2 above)
Comment (dc03-work) [reply llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp]: Done.

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp (hunk 3 above)
Comment (dc03-work) [reply llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp]: Done.

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp (hunk 5 above)
Comment (dc03-work) [reply llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp]: Okay, I have updated the patch to use this call.

**Comment by madhur13490** [comment]
LGTM

**Review by madhur13490 — APPROVED** [review]

**Pull request merged**